velox/core/QueryCtx.cpp

```diff
@@ -91,4 +120,99 @@ void QueryCtx::updateSpilledBytesAndCheckLimit(uint64_t bytes) {
   }
 }
 
+std::unique_ptr<memory::MemoryReclaimer> QueryCtx::MemoryReclaimer::create(
+    QueryCtx* queryCtx,
+    memory::MemoryPool* pool) {
+  return std::unique_ptr<memory::MemoryReclaimer>(
+      new QueryCtx::MemoryReclaimer(queryCtx->shared_from_this(), pool));
+}
+
+uint64_t QueryCtx::MemoryReclaimer::reclaim(
+    memory::MemoryPool* pool,
+    uint64_t targetBytes,
+    uint64_t maxWaitMs,
+    memory::MemoryReclaimer::Stats& stats) {
+  auto queryCtx = ensureQueryCtx();
+  if (queryCtx == nullptr) {
+    return 0;
+  }
+  VELOX_CHECK_EQ(pool->name(), pool_->name());
+
+  const auto leaveGuard =
+      folly::makeGuard([&]() { queryCtx->finishArbitration(); });
+  queryCtx->startArbitration();
+  // Sort the child pools based on their reserved memory and reclaim from the
+  // child pool with most reservation first.
+  struct Candidate {
+    std::shared_ptr<memory::MemoryPool> pool;
+    int64_t reclaimableBytes;
+  };
+  std::vector<Candidate> candidates;
+  {
+    std::shared_lock guard{pool->poolMutex_};
+    candidates.reserve(pool->children_.size());
+    for (auto& entry : pool->children_) {
+      auto child = entry.second.lock();
+      if (child != nullptr) {
+        const int64_t reclaimableBytes = child->reclaimableBytes().value_or(0);
+        candidates.push_back(Candidate{std::move(child), reclaimableBytes});
+      }
+    }
+  }
+
+  std::sort(
+      candidates.begin(),
+      candidates.end(),
+      [](const auto& lhs, const auto& rhs) {
+        return lhs.reclaimableBytes > rhs.reclaimableBytes;
+      });
+
+  uint64_t reclaimedBytes{0};
+  for (const auto& candidate : candidates) {
+    if (candidate.reclaimableBytes == 0) {
+      break;
+    }
+    const auto bytes = candidate.pool->reclaim(targetBytes, maxWaitMs, stats);
+    reclaimedBytes += bytes;
```

Comment:
other than additional preparations at the top, there isn't too much difference from memory::MemoryReclaimer? Can we 
...do prep...
memory::MemoryReclaimer::reclaim();

Reply:
Good point. We check on the reclaimable bytes instead of capacity, and I will move this into memory reclaimer in followup.